File: tool/src/main/java/org/kravemir/svg/labels/tool/project/MakefileBuilder.java
package org.kravemir.svg.labels.tool.project;

import java.util.stream.Collectors;
import java.util.stream.Stream;

public class MakefileBuilder {

    private StringBuilder target;

    public MakefileBuilder() {
        this.target = new StringBuilder();
    }

    public void appendVariable(String name, String value) {
        appendVariable(name, value, false);
    }

    public void appendVariable(String name, String value, boolean export) {
        target.append(name);
        target.append(" := ");
        target.append(value);
        target.append("\n");

        if(export) {
            target.append("export ");
            target.append(name);
            target.append("\n");
        }
    }

    public void appendRule(String output, Stream<String> depends, String... commands) {
        appendRule(
                output,
                depends.collect(Collectors.joining(" ")),
                commands
        );
    }

    public void appendRule(String output, String depends, String... commands) {
        target.append(output);
        target.append(": ");
        target.append(depends);

        for(String cmd : commands) {
            target.append("\n\t");
            target.append(cmd);
        }

        target.append("\n\n");
    }

    public void appendMkdirRule(String dirname) {
        target.append(dirname);
        target.append(": ;\n\tmkdir -p $@\n\n");
    }

    public void append(CharSequence csq) {
        target.append(csq);
    }

    @Override
    public String toString() {
        return target.toString();
    }
}

File: internal/project/makefile_builder.go
package project

import (
	"fmt"
	"strings"
)

type makefileBuilder struct {
	builder strings.Builder
}

func (b *makefileBuilder) appendVariable(name string, value string) {
	b.builder.WriteString(name)
	b.builder.WriteString(" := ")
	b.builder.WriteString(value)
	b.builder.WriteString("\n")
}

func (b *makefileBuilder) appendVariableWithExport(name string, value string) {
	b.appendVariable(name, value)

	b.builder.WriteString("export ")
	b.builder.WriteString(name)
	b.builder.WriteString("\n")
}

func (b *makefileBuilder) appendRule(output string, depends string, commands ...string) {
	b.builder.WriteString(output)
	b.builder.WriteString(":")
	if depends != "" {
		b.builder.WriteString(" " + depends)
	}

	for _, cmd := range commands {
		b.builder.WriteString("\n\t")
		b.builder.WriteString(cmd)
	}

	b.builder.WriteString("\n\n")
}

func (b *makefileBuilder) appendMkdirRule(dirname string) {
	b.builder.WriteString(dirname)
	b.builder.WriteString(": ;\n\tmkdir -p $@\n\n")
}

func (b *makefileBuilder) appendInclude(path string) {
	b.builder.WriteString(fmt.Sprintf("include %s\n", path))
}

func (b *makefileBuilder) append(s string) {
	b.builder.WriteString(s)
}
